### PR TITLE
Backport current cloud hypervisor vfio patches to external vfio-ioctls crate

### PR DESCRIPTION
**Pull request by arronwy**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -8,11 +8,11 @@ license = "Apache-2.0 OR BSD-3-Clause"
 edition = "2018"
 
 [dependencies]
-byteorder = "~1.2.1"
+byteorder = "~1.3.1"
 log = "0.4"
-kvm-bindings = "~0"
-kvm-ioctls = "~0"
+kvm-bindings = { git = "https://github.com/cloud-hypervisor/kvm-bindings", branch = "ch" }
```
Comment (jiangliu) [inline Cargo.toml:13]: Why there's special dependency on the kvm-bindings/kvm-ioclts?

File: src/dma_mapping.rs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+use std::io;
+use std::sync::Arc;
+use vm_memory::{GuestAddress, GuestAddressSpace, GuestMemoryAtomic, GuestMemoryMmap};
+
+/// This structure implements the ExternalDmaMapping trait. It is meant to
+/// be used when the caller tries to provide a way to update the mappings
+/// associated with a specific VFIO container.
+pub struct VfioDmaMapping {
+    container: Arc<VfioContainer>,
+    memory: GuestMemoryAtomic<GuestMemoryMmap>,
```
Comment (jiangliu) [inline src/dma_mapping.rs]: Is it ok to use generic type <M: GuestMemory > instead of GuestMemoryMmap here?

File: src/vfio_device.rs [hunk 3]
```diff
@@ -21,7 +21,7 @@ use kvm_bindings::{
 use kvm_ioctls::DeviceFd;
 use log::{debug, error, warn};
 use vfio_bindings::bindings::vfio::*;
-use vm_memory::{Address, GuestMemory, GuestMemoryRegion, MemoryRegionAddress};
+use vm_memory::{Address, GuestMemory, GuestMemoryRegion, GuestMemoryMmap, GuestRegionMmap, MemoryRegionAddress};
```
Comment (jiangliu) [inline src/vfio_device.rs]: We should use generic type <M: GuestMemory> instead of concrete GuestMemoryMmap here.

File: src/vfio_device.rs [hunk 4]
```diff
@@ -636,7 +637,11 @@ impl VfioDevice {
     /// # Parameters
     /// * `sysfspath`: specify the vfio device path in sys file system.
     /// * `container`: the new VFIO device object will bind to this container object.
```
Comment (jiangliu) [inline src/vfio_device.rs:639]: Please add doc for iommu_attached:)

File: src/vfio_device.rs [hunk 5]
```diff
@@ -933,6 +939,57 @@ impl VfioDevice {
...
+            })?;
+        }
+        Ok(())
+    }
+
+    /// Update the vfio device container's iommu table when guest memory is extended.
+    ///
+    /// # Parameters
+    /// * new_region: extended guest memory region which should be updated.
+    pub fn extend_dma_map(&self, new_region: &Arc<GuestRegionMmap>) -> Result<()> {
```
Comment (jiangliu) [inline src/vfio_device.rs:985]: Do we need shrink_dma_map() to undo what's done by extend_dma_map()?

File: Cargo.toml (hunk 1 above)
Comment (arronwy) [reply Cargo.toml:13]: Cloud Hypervisor have dependencies for the KVM serialization work,  kvm-bindings/kvm-ioclts related patches will submit to upstream, after that we can use the same base, else we will have building error due to different versions.

File: src/vfio_device.rs (hunk 3 above)
Comment (arronwy) [reply src/vfio_device.rs]: In Cloud Hypervisor, we switched to use mem: GuestMemoryAtomic< GuestMemoryMmap > to save memory, we found Cloud Hypervisor will have build error if use generic type <M: GuestMemory> in vfio-ioctls:
```
.setup_dma_map(&self.mem.memory())
    ^^^^^^^^^^ the trait `vm_memory::GuestMemory` is not implemented for `vm_memory::GuestMemoryLoadGuard<vm_memory::GuestMemoryMmap>`
```
Any suggestions?

File: src/vfio_device.rs (hunk 4 above)
Comment (arronwy) [reply src/vfio_device.rs:639]: Ok, will add it, thanks.

File: src/vfio_device.rs (hunk 3 above)
Comment (jiangliu) [reply src/vfio_device.rs]: Could you please try .setup_dma_map(self.mem.memory().deref())?

Comment (arronwy) [reply src/vfio_device.rs]: it works, thanks :-)

File: Cargo.toml [hunk 6]
```diff
@@ -8,11 +8,11 @@ license = "Apache-2.0 OR BSD-3-Clause"
...
+kvm-ioctls = { git = "https://github.com/cloud-hypervisor/kvm-ioctls", branch = "ch" }
 vfio-bindings = "~0"
 vmm-sys-util = "~0"
 
-vm-memory = { git = "https://github.com/rust-vmm/vm-memory", features = ["backend-mmap"] }
+vm-memory = { version = "0.2.0" , features = ["backend-mmap", "backend-atomic"] }
```
Comment (jiangliu) [inline Cargo.toml]: How about "version = "^0, >=0.2", please refer https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html

File: src/dma_mapping.rs (hunk 2 above)
Comment (jiangliu) [reply src/dma_mapping.rs]: Seems we should use "AS: GuestAddressSpace" to replace "GuestMemoryAtomic<GuestMemoryMmap>" here.

File: src/dma_mapping.rs [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    /// * `memory·: guest memory to mmap.
+    pub fn new(container: Arc<VfioContainer>, memory: GuestMemoryAtomic<GuestMemoryMmap>) -> Self {
+        VfioDmaMapping { container, memory }
+    }
+}
+
+impl ExternalDmaMapping for VfioDmaMapping {
+    fn map(&self, iova: u64, gpa: u64, size: u64) -> std::result::Result<(), io::Error> {
+        let user_addr = if let Some(addr) = get_host_address_range(
+            &self.memory.memory(),
```
Comment (jiangliu) [inline src/dma_mapping.rs]: This may fail to support memory hot-add, but we may treat it as a bug and fix it later.

File: Cargo.toml (hunk 6 above)
Comment (arronwy) [reply Cargo.toml]: Yes, it makes sense, thanks :)

File: src/vfio_device.rs (hunk 5 above)
Comment (arronwy) [reply src/vfio_device.rs:985]: Current Cloud Hypervisor only support memory hotplug, for unplug operation, it only takes effect after vm restarted.  I not sure whether we need add the shrink API here?

File: src/dma_mapping.rs (hunk 2 above)
Comment (arronwy) [reply src/dma_mapping.rs]: We found use generic type memory which require get_host_address_range/get_region_host_address_range  all switch to the generit type, it makes some operations more complicate to convert.

Comment (jiangliu) [reply src/dma_mapping.rs]: That should be the way, otherwise it may not be reused by other projects:)

Comment (arronwy) [reply src/dma_mapping.rs]: > GuestAddressSpace

we can use memory: Arc < GuestMemory > , did you mean we use memory: Arc < GuestAddressSpace >  or just memory: GuestAddressSpace?

File: Cargo.toml (hunk 1 above)
Comment (sameo) [reply Cargo.toml:13]: @arronwy fwiw, we don't need the cloud-hypervisor version of kvm-bindings to submit this upstream.

**Review by sameo — APPROVED** [review]

**Review by jiangliu — APPROVED** [review]

File: Cargo.toml (hunk 1 above)
Comment (arronwy) [reply Cargo.toml:13]: I found if we use the upstream kvm-binding, cloud-hypervisor will failed to build with following error:
```
error[E0308]: mismatched types
   --> vfio-ioctls/src/vfio_device.rs:200:30
    |
200 |  .set_device_attr(&dev_attr)
    |   ^^^^^^^^^ expected struct `kvm_bindings::x86::bindings_v4_20_0::kvm_device_attr`, found struct `kvm_bindings::kvm_device_attr`
    |
    = note: expected reference `&kvm_bindings::x86::bindings_v4_20_0::kvm_device_attr`
               found reference `&kvm_bindings::kvm_device_attr`
    = note: perhaps two different versions of crate `kvm_bindings` are being used?
```

**Comment by sameo** [comment]
@sboeuf Could you please have a look? If it looks fine to you, we can merge and switch the CLH master branch to it.

File: src/vfio_device.rs [hunk 8]
```diff
@@ -628,6 +628,7 @@ pub struct VfioDevice {
     irqs: HashMap<u32, VfioIrq>,
     group: Arc<VfioGroup>,
     container: Arc<VfioContainer>,
+    iommu_attached: bool,
```
Comment (sboeuf) [inline src/vfio_device.rs:631]: The introduction of this field should be part of a separate commit.
First you should introduce the DMA setup wrappers, and then introduce the special handling when it is attached to a virtual IOMMU.

File: src/vfio_device.rs (hunk 5 above)
Comment (sboeuf) [reply src/vfio_device.rs:985]: Implementing `shrink_dma_map()` would be nice but it's not needed.
With memory hotplug through ACPI, because the unplug is not supported, there is no need for a shrink API.
With `virtio-mem` support, unplug is supported but `virtio-mem` uses one dedicated region defined upfront for ballooning purposes. Whenever the first hotplug happens, the entire region is hotplugged. But later whenever a new hotplug or unplug happens, it's always part of this region, which means VFIO map does not need to be updated.
I definitely agree we could try to apply some shrinking in order to control at a finer granularity than the region, but this would involve further changes about the overall design. So might be worth keeping this in mind and implementing it later :)

File: src/vfio_device.rs (hunk 8 above)
Comment (arronwy) [reply src/vfio_device.rs:631]: Seperated now, thanks:)

Comment (sboeuf) [reply src/vfio_device.rs:631]: Thanks @arronwy

**Review by sboeuf — APPROVED** [review]
